File: solutions/0852. Peak Index in a Mountain Array/0852_2.cpp
// Time complexity: O(log⁡n)
// Space complexity: O(1)
class Solution {
public:
    int peakIndexInMountainArray(vector<int>& arr) {
        int start = 0, end = arr.size() - 1, mid;

        while (start < end) {
            mid = (start + end) / 2;
            if (arr[mid] < arr[mid + 1]) 
                start = mid + 1;
            else
                end = mid;
        }
        return start;
    }
};

File: solutions/0852. Peak Index in a Mountain Array/0852_2.py
# Time complexity: O(log⁡n)
# Space complexity: O(1)
class Solution :
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        start = 0
        end = len(arr)

        while start < end:
            mid = (start + end) // 2
            if arr[mid] < arr[mid + 1]: 
                start = mid + 1
            else:
                end = mid

        return start